File: iosApp/iosApp/Features/ComfyUI/SDWebUIGenerationViewModel.swift
import SwiftUI
import Shared
import UIKit

@MainActor
final class SDWebUIGenerationViewModelOwner: ObservableObject {
    let vm: Feature_comfyuiSDWebUIGenerationViewModel
    private let store = ViewModelStore()

    @Published var models: [String] = []
    @Published var samplers: [String] = []
    @Published var selectedModel: String = ""
    @Published var selectedSampler: String = "Euler"
    @Published var prompt: String = ""
    @Published var negativePrompt: String = ""
    @Published var steps: Double = 20
    @Published var cfgScale: Double = 7.0
    @Published var width: Int32 = 512
    @Published var height: Int32 = 512
    @Published var seed: String = "-1"
    @Published var isLoading: Bool = false
    @Published var isGenerating: Bool = false
    @Published var progress: Double = 0
    @Published var progressStep: Int32 = 0
    @Published var progressTotalSteps: Int32 = 0
    @Published var generatedImages: [UIImage] = []
    @Published var error: String?

    init() {
        vm = KoinHelper.shared.createSDWebUIGenerationViewModel()
        store.put(key: "SDWebUIGenerationViewModel", viewModel: vm)
    }

    deinit { store.clear() }

    func observeUiState() async {
        for await state in vm.uiState {
            models = state.models as? [String] ?? []
            samplers = state.samplers as? [String] ?? []
            selectedModel = state.selectedModel
            selectedSampler = state.selectedSampler
            isLoading = state.isLoading
            isGenerating = state.isGenerating
            progress = state.progress
            progressStep = state.progressStep
            progressTotalSteps = state.progressTotalSteps
            let base64List = state.generatedImages as? [String] ?? []
            generatedImages = base64List.compactMap { b64 in
                guard let data = Data(base64Encoded: b64) else { return nil }
                return UIImage(data: data)
            }
            error = state.error
        }
    }

    var isSeedValid: Bool { parsedSeed(seed) != nil }

    /// KMP `onGenerate()` ignores a blank prompt, and an unparseable seed never reaches KMP.
    var canGenerate: Bool {
        !prompt.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && isSeedValid
    }

    func onPromptChanged(_ value: String) {
        prompt = value
        vm.onPromptChanged(value: value)
    }
    func onNegativePromptChanged(_ value: String) {
        negativePrompt = value
        vm.onNegativePromptChanged(value: value)
    }
    func onModelSelected(_ model: String) {
        selectedModel = model
        vm.onModelSelected(model: model)
    }
    func onSamplerSelected(_ sampler: String) {
        selectedSampler = sampler
        vm.onSamplerSelected(sampler: sampler)
    }
    /// Rounds so the slider, its label and the submitted step count stay the same integer.
    func onStepsChanged(_ steps: Double) {
        let rounded = steps.rounded()
        self.steps = rounded
        vm.onStepsChanged(steps: Int32(rounded))
    }
    func onCfgChanged(_ cfg: Double) {
        cfgScale = cfg
        vm.onCfgChanged(cfg: cfg)
    }
    func onWidthChanged(_ width: Int32) {
        self.width = width
        vm.onWidthChanged(w: width)
    }
    func onHeightChanged(_ height: Int32) {
        self.height = height
        vm.onHeightChanged(h: height)
    }
    /// Blank text means a random seed. Unparseable text stays in the field without reaching KMP.
    func onSeedChanged(_ text: String) {
        seed = text
        if let value = parsedSeed(text) { vm.onSeedChanged(seed: value) }
    }
    func onDismissError() {
        error = nil
        vm.onDismissError()
    }
    func onGenerate() { vm.onGenerate() }
    func onInterrupt() { vm.onInterrupt() }
}

/// KMP seed value meaning "random"; an empty seed field shows it as the placeholder.
private let randomSeed: Int64 = -1

private func parsedSeed(_ text: String) -> Int64? {
    let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
    return trimmed.isEmpty ? randomSeed : Int64(trimmed)
}

File: feature/feature-comfyui/src/commonMain/kotlin/com/riox432/civitdeck/feature/comfyui/presentation/SDWebUIGenerationViewModel.kt
package com.riox432.civitdeck.feature.comfyui.presentation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.riox432.civitdeck.domain.model.SDWebUIGenerationParams
import com.riox432.civitdeck.domain.model.SDWebUIGenerationProgress
import com.riox432.civitdeck.domain.util.suspendRunCatching
import com.riox432.civitdeck.feature.comfyui.domain.usecase.FetchSDWebUIModelsUseCase
import com.riox432.civitdeck.feature.comfyui.domain.usecase.FetchSDWebUISamplersUseCase
import com.riox432.civitdeck.feature.comfyui.domain.usecase.FetchSDWebUIVaesUseCase
import com.riox432.civitdeck.feature.comfyui.domain.usecase.GenerateSDWebUIImageUseCase
import com.riox432.civitdeck.feature.comfyui.domain.usecase.InterruptSDWebUIGenerationUseCase
import com.riox432.civitdeck.util.Logger
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class SDWebUIGenerationUiState(
    val models: List<String> = emptyList(),
    val samplers: List<String> = emptyList(),
    val vaes: List<String> = emptyList(),
    val selectedModel: String = "",
    val selectedSampler: String = "Euler",
    val selectedVae: String = "",
    val prompt: String = "",
    val negativePrompt: String = "",
    val steps: Int = 20,
    val cfgScale: Double = 7.0,
    val width: Int = 512,
    val height: Int = 512,
    val seed: Long = -1,
    val isLoading: Boolean = false,
    val isGenerating: Boolean = false,
    val progress: Double = 0.0,
    val progressStep: Int = 0,
    val progressTotalSteps: Int = 0,
    val generatedImages: List<String> = emptyList(),
    val error: String? = null,
)

class SDWebUIGenerationViewModel(
    private val fetchModels: FetchSDWebUIModelsUseCase,
    private val fetchSamplers: FetchSDWebUISamplersUseCase,
    private val fetchVaes: FetchSDWebUIVaesUseCase,
    private val generateImage: GenerateSDWebUIImageUseCase,
    private val interruptGeneration: InterruptSDWebUIGenerationUseCase,
) : ViewModel() {

    private val _uiState = MutableStateFlow(SDWebUIGenerationUiState())
    val uiState: StateFlow<SDWebUIGenerationUiState> = _uiState.asStateFlow()

    private var generationJob: Job? = null

    init { loadResources() }

    private fun loadResources() {
        _uiState.update { it.copy(isLoading = true) }
        viewModelScope.launch {
            try {
                val models = suspendRunCatching { fetchModels() }.getOrDefault(emptyList())
                val samplers = suspendRunCatching { fetchSamplers() }.getOrDefault(emptyList())
                val vaes = suspendRunCatching { fetchVaes() }.getOrDefault(emptyList())
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        models = models,
                        samplers = samplers,
                        vaes = vaes,
                        selectedModel = models.firstOrNull() ?: "",
                        selectedSampler = samplers.firstOrNull() ?: "Euler",
                    )
                }
            } catch (@Suppress("TooGenericExceptionCaught") e: Exception) {
                Logger.e(TAG, "Failed to load resources: ${e.message}")
                _uiState.update { it.copy(isLoading = false, error = e.message) }
            }
        }
    }

    fun onPromptChanged(value: String) { _uiState.update { it.copy(prompt = value) } }
    fun onNegativePromptChanged(value: String) { _uiState.update { it.copy(negativePrompt = value) } }
    fun onModelSelected(model: String) { _uiState.update { it.copy(selectedModel = model) } }
    fun onSamplerSelected(sampler: String) { _uiState.update { it.copy(selectedSampler = sampler) } }
    fun onStepsChanged(steps: Int) { _uiState.update { it.copy(steps = steps) } }
    fun onCfgChanged(cfg: Double) { _uiState.update { it.copy(cfgScale = cfg) } }
    fun onWidthChanged(w: Int) { _uiState.update { it.copy(width = w) } }
    fun onHeightChanged(h: Int) { _uiState.update { it.copy(height = h) } }
    fun onSeedChanged(seed: Long) { _uiState.update { it.copy(seed = seed) } }
    fun onDismissError() { _uiState.update { it.copy(error = null) } }

    fun onGenerate() {
        val state = _uiState.value
        if (state.isGenerating || state.prompt.isBlank()) return
        _uiState.update {
            it.copy(isGenerating = true, generatedImages = emptyList(), error = null, progress = 0.0)
        }
        val params = SDWebUIGenerationParams(
            prompt = state.prompt,
            negativePrompt = state.negativePrompt,
            steps = state.steps,
            cfgScale = state.cfgScale,
            width = state.width,
            height = state.height,
            samplerName = state.selectedSampler,
            seed = state.seed,
        )
        generationJob = viewModelScope.launch {
            generateImage(params).collect { progress ->
                handleProgress(progress)
            }
        }
    }

    private fun handleProgress(progress: SDWebUIGenerationProgress) {
        when (progress) {
            is SDWebUIGenerationProgress.Generating -> _uiState.update {
                it.copy(
                    progress = progress.fraction,
                    progressStep = progress.step,
                    progressTotalSteps = progress.totalSteps,
                )
            }
            is SDWebUIGenerationProgress.Completed -> _uiState.update {
                it.copy(isGenerating = false, generatedImages = progress.base64Images, progress = 1.0)
            }
            is SDWebUIGenerationProgress.Error -> _uiState.update {
                it.copy(isGenerating = false, error = progress.message)
            }
        }
    }

    fun onInterrupt() {
        generationJob?.cancel()
        viewModelScope.launch {
            suspendRunCatching { interruptGeneration() }
                .onFailure { e -> Logger.w(TAG, "Interrupt failed: ${e.message}") }
            _uiState.update { it.copy(isGenerating = false) }
        }
    }

    companion object {
        private const val TAG = "SDWebUIGenerationVM"
    }
}
